Add podman builder to support CRI-O runtime

mochiliu3000 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Environments like OCP 4 use CRI-O as default container runtime. In order to submit jobs to such environment, need to support podman builder to build image and push image to CRI-O based registry with/without TLS checking.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #472 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note

```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/fairing/474)
<!-- Reviewable:end -->

jinchihe commented:
/ok-to-test

jinchihe commented:
@mochiliu3000 Please fix the beflow lint problem in `kubeflow/fairing/builders/podman/podman.py`, thanks.
```
INFO|2020-03-06T06:54:05|util.py:72| ************* Module kubeflow.fairing.builders.podman.podman
INFO|2020-03-06T06:54:05|util.py:72| C: 44, 0: Line too long (110/100) (line-too-long)
INFO|2020-03-06T06:54:05|util.py:72| C: 64, 0: Line too long (102/100) (line-too-long)
INFO|2020-03-06T06:54:05|util.py:72| C: 74, 0: Line too long (109/100) (line-too-long)
INFO|2020-03-06T06:54:05|util.py:72| C: 80, 0: Line too long (104/100) (line-too-long)
INFO|2020-03-06T06:54:05|util.py:72| C: 83, 0: Final newline missing (missing-final-newline)
INFO|2020-03-06T06:54:05|util.py:72| W:  1, 0: Unused import json (unused-import)
```

jinchihe commented on examples/kubeflow/main_ocp.py:
```diff
@@ -0,0 +1,84 @@
+# Copyright 2015 The TensorFlow Authors. All Rights Reserved.
```
`2015 The TensorFlow Authors` ->  `2020 The Kubeflow Authors `

Please also add the License information to new Python script in the PR. Thanks

jinchihe commented on kubeflow/fairing/builders/podman/podman.py:
```diff
@@ -0,0 +1,83 @@
+            base_image=base_image,
+            preprocessor=preprocessor,
+            dockerfile_path=dockerfile_path)
+        self.tls_verify = tls_verify
```
Would you please add defailed description for the parameters? the format can be as below, so that we can update the Fairing API docs for the builder later. please add docString for all function and Classes, Thanks.
```
        ""Description for the function....

        :param registry: basic description ...
        :param image_name: basic description ...

    
        """
```

jinchihe commented on kubeflow/fairing/builders/podman/podman.py:
```diff
@@ -0,0 +1,83 @@
+        logger.warning('Building podman image {}...'.format(self.image_tag))
+
+        # Due to this issue, instead of using 'podman_client.images.build', call command line to build
+        # https://github.com/containers/python-podman/issues/51
```
I suggest add a TBD here as beflow, once the issue fixed we can use the SDK Client, instead of command line.
```
#TBD @mochiliu3000 Due to this issue...
```

jinchihe commented on kubeflow/fairing/builders/podman/podman.py:
```diff
@@ -0,0 +1,83 @@
+        cmd_push = 'podman push {} --tls-verify={}'.format(self.image_tag, str(self.tls_verify).lower())
+        push_return = os.system(cmd_push)
+        if push_return != 0:
+            raise Exception('Image push failed')
```
Add new line here please

mochiliu3000 commented:
@jinchihe Follow your requirements, I've updated the PR and also added a test case: test_podman.py. Please help review, thanks.

<img width="1428" alt="屏幕快照 2020-03-24 下午8 28 29" src="https://user-images.githubusercontent.com/8138679/77425773-366e0980-6e0e-11ea-926a-4baf5f74611f.png">

jinchihe commented:
/lgtm
/approve
